homeassistant/components/dio_chacon/cover.py

```diff
@@ -0,0 +1,164 @@
+
+from . import DioChaconData
+from .entity import DioChaconEntity
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Defer sensor setup to the shared sensor module."""
+    dio_chacon_data: DioChaconData = config_entry.runtime_data
+
+    entities: list[DioChaconCover] = []
+    for device in dio_chacon_data.list_devices:
+        if device["type"] == DeviceTypeEnum.SHUTTER.value:
+            cover = DioChaconCover(dio_chacon_data.dio_chacon_client, device)
+            # Registers a callback to update the device state when the server sends an event
+            dio_chacon_data.dio_chacon_client.set_callback_device_state_by_device(
+                cover.target_id, cover.callback_device_state
+            )
+            entities.append(cover)
+
+    async_add_entities(entities)
+
+
+class DioChaconCover(DioChaconEntity, CoverEntity):
+    """Object for controlling a Dio Chacon cover."""
+
+    # To avoid having name twice
+    _attr_has_entity_name = False
+
+    _attr_device_class = CoverDeviceClass.SHUTTER
+
+    _attr_supported_features = (
+        CoverEntityFeature.OPEN
+        | CoverEntityFeature.CLOSE
+        | CoverEntityFeature.STOP
+        | CoverEntityFeature.SET_POSITION
+    )
+
+    def __init__(
+        self,
+        dio_chacon_client: DIOChaconAPIClient,
+        device: dict[str, Any],
+    ) -> None:
+        """Initialize the cover."""
+
+        _LOGGER.debug(
+            "Adding DIO Chacon SHUTTER Cover with id=%s, name=%s, openlevel=%s, movement=%s and connected=%s",
+            device["id"],
+            device["name"],
+            device["openlevel"],
+            device["movement"],
+            device["connected"],
+        )
+
+        super().__init__(
+            dio_chacon_client, device["id"], device["name"], device["model"]
+        )
+
+        self._attr_available = device["connected"]
+        self._attr_current_cover_position = device["openlevel"]
+        self._attr_is_closing = device["movement"] == ShutterMoveEnum.DOWN.value
+        self._attr_is_opening = device["movement"] == ShutterMoveEnum.UP.value
+        self._attr_is_closed = self._attr_current_cover_position == 0
+
+    # callback coroutine for device state notification pushed from the server
+    def callback_device_state(self, data: dict[str, Any]) -> None:
+        """Receive callback for device state notification pushed from the server."""
+
+        _LOGGER.debug("Data received from server %s", data)
+        self._attr_available = data["connected"]
+        self._attr_current_cover_position = data["openlevel"]
+        self._attr_is_closing = data["movement"] == ShutterMoveEnum.DOWN.value
+        self._attr_is_opening = data["movement"] == ShutterMoveEnum.UP.value
+        self._attr_is_closed = self._attr_current_cover_position == 0
+        self.async_write_ha_state()
```

Comment:
```suggestion
    def callback_device_state(self, data: dict[str, Any]) -> None:
        """Receive callback for device state notification pushed from the server."""

        _LOGGER.debug("Data received from server %s", data)
        self._update_attr(data)
        self.async_write_ha_state()
```

I think this might even work the same as with the switches you're going to add, so maybe its a nice idea to move the `_update_attr` as a function to the `entity.py`.

That would just look like
```py
def _update_attr(self) -> None:
    """Update attributes."""
```

And then just add the code suggestion I put on top of this comment in the base class, then you don't have to worry about the callback in the platform classes anymore.

(This might be a bit confusing, feel free to shoot questions back if you are unsure :) )

Reply:
Done